portmapper/proxy.go

```diff
@@ -77,21 +77,28 @@ func (p *proxyCommand) Stop() error {
 // port allocations on bound port, because without userland proxy we using
 // iptables rules and not net.Listen
 type dummyProxy struct {
-	listener io.Closer
-	addr     net.Addr
+	listener  io.Closer
+	addr      net.Addr
+	ipVersion int
 }
 
 func newDummyProxy(proto string, hostIP net.IP, hostPort int) (userlandProxy, error) {
+	// detect version of hostIP to bind only to correct version
+	version := 6
+	if hostIP.To4() != nil {
+		version = 4
+	}
+
 	switch proto {
 	case "tcp":
 		addr := &net.TCPAddr{IP: hostIP, Port: hostPort}
-		return &dummyProxy{addr: addr}, nil
+		return &dummyProxy{addr: addr, ipVersion: version}, nil
```

Comment:
this might break backward compatibility
trying to understand why this func gets called twice when `ip6tables` is enabled

Reply:
The interface should not be a problem because I only changed a private part but the behavior changed a bit.

Until now a `net.ListenTCP("tcp", addr)` was used which caused a listen on IPv4 and IPv6. Because the port allocation is now done separate it is also calling the proxy two times. This causes a listen on `0.0.0.0` and on `::` which will collide if the proto is `tcp`.

I know that this kind of change would be a breaking change for the real proxy. But for the dummy it only has the effect that the IPv4 port is blocked for a `0.0.0.0` instead of both. For specific host IP bindings (like `-p 192.168.0.42:80:80`) there should not be any changes.

Reply:
thanks for clarifying that, this snippet now makes sense
looks like the double call is due to https://github.com/moby/libnetwork/blob/9f028a109a91a8274809e64c7d274fc11483063c/drivers/bridge/port_mapping.go#L19
since if the container has a dual stack, and when you publish using`-p 8080:8080`, it should bind on both IPv6 and IPv4

also, shouldn't we be passing the true value of `ulPxyEnabled` here instead of `false`, https://github.com/moby/libnetwork/blob/5c6a95bfb20c61571a00f913c6b91959ede84e8d/drivers/bridge/port_mapping.go#L38

Reply:
we should also skip this code path when binding to a specific ipv4 addr /  `-p 192.168.0.42:80:80` https://github.com/moby/libnetwork/blob/9f028a109a91a8274809e64c7d274fc11483063c/drivers/bridge/port_mapping.go#L36

Reply:
I think I set the `ulPxyEnabled` to false in 
https://github.com/moby/libnetwork/blob/5c6a95bfb20c61571a00f913c6b91959ede84e8d/drivers/bridge/port_mapping.go#L38
because I had already trouble with the proxy in the past.

Based on this lines the change commented above should be
```go
if config.EnableIP6Tables && config.EnableUserlandProxy {
```
istead of 
```go
if config.EnableIPTables && config.EnableIP6Tables && config.EnableUserlandProxy {
```
because the first `allocatePortsInternal` is called regardless of the `config.EnableIPTables`.

So in the end It is a simple: If the `ip6tables` are enabled the userland proxy has to be disabled.

Reply:
> we should also skip this code path when binding to a specific ipv4 addr e.g. `-p 192.168.0.42:80:80`
> 
> https://github.com/moby/libnetwork/blob/9f028a109a91a8274809e64c7d274fc11483063c/drivers/bridge/port_mapping.go#L36

This should be handled via 9f028a109a91a8274809e64c7d274fc11483063c. This basically is doing exactly this and also the same with IPv6 bindings. 
* `-p 192.168.0.42:80:80` would skip IPv6
* `-p [fd00:dead:beef::42]:80:80` would skip IPv4

Based on container and host IP version mismatch. But also only if `ip6tables` is enabled because in the past it was done for both IP versions and I wont break anything.